### PR TITLE
[fix] ActiveStorage Registry failing when service is a Proc

### PR DESCRIPTION
**Pull request by swiknaba**
### Motivation / Background

PR https://github.com/rails/rails/pull/49863 introduced that `service` for ActiveRecord callbacks `has_one_attached`/`has_many_attached` can be a `proc`.

Thanks for that feature; I love it. Unfortunately, it fails for me with

> activestorage-7.1.2/lib/active_storage/service/registry.rb:11
> `undefined method `to_sym' for #<Proc:0x00000001219b`
>
> services.fetch(name.to_sym) do |key|

This PR fixes this :)

### Detail

`ActiveStorage::Service::Registry#fetch` now checks if the passed `service` is a Proc.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] ~Tests are added or updated if you fix a bug or add a feature.~
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activestorage/lib/active_storage/service/registry.rb [hunk 1]
```diff
@@ -7,7 +7,8 @@ def initialize(configurations)
       @services = {}
     end
 
-    def fetch(name)
+    def fetch(service)
+      name = service.is_a?(Proc) ? service.call : service
```
Comment (swiknaba) [inline activestorage/lib/active_storage/service/registry.rb:11]: I'm not passing any arguments for my usecase. Do we have `record` magically available here? We probably should allow passing an argument to `service.call` here.

Comment (swiknaba) [reply activestorage/lib/active_storage/service/registry.rb]: given the original implementation, it seems we *must* pass the record to the services proc, so just adding it in here as well.

**Comment by jonathanhefner** [comment]
Thank you for the pull request.  However, the error you encountered is because the feature has not been released yet.  See https://github.com/rails/rails/pull/49863#issuecomment-1848681462.

**Comment by swepp** [comment]
> Thank you for the pull request. However, the error you encountered is because the feature has not been released yet. See [#49863 (comment)](https://github.com/rails/rails/pull/49863#issuecomment-1848681462).

@jonathanhefner  Shouldn't we fix a broken code path, nonetheless?

**Comment by swiknaba** [comment]
> Thank you for the pull request. However, the error you encountered is because the feature has not been released yet. See [#49863 (comment)](https://github.com/rails/rails/pull/49863#issuecomment-1848681462).

@jonathanhefner  Shouldn't we fix a broken code path, nonetheless?

**Comment by swiknaba** [comment]
> Thank you for the pull request. However, the error you encountered is because the feature has not been released yet. See [#49863 (comment)](https://github.com/rails/rails/pull/49863#issuecomment-1848681462).

That codepath is broken nonetheless, isn't it?

**Comment by jonathanhefner** [comment]
The error you encountered should be prevented by conditionals like [this one](https://github.com/rails/rails/pull/49863/files#diff-50646cbd5d476624e64996685b0256c97cdd46b742ab8d3b80adce4e9663b39aR101).  Are you able to reproduce the error on Rails `main` branch?

**Pull request closed without merging**